File: src_web3/phase0/web3_vs_web2_concept.ts
import { createHash } from "crypto"; // We import createHash to generate content-addressable identifiers for Web3 data integrity proofs.

// PHASE 0: Web3 vs Web2 Concept in TypeScript
// This file demonstrates the paradigm shift using Node.js types and hashing.

/// Represents a Web2 platform API where a single company owns all state.
/// We use a class because Web2 platforms encapsulate hidden business logic.
class Web2Platform {
    private owner: string; // The owner field stores the entity that can unilaterally change rules or ban users.
    private userData: Map<string, string>; // A private Map simulates the hidden corporate database users cannot audit.
    private revenueSharePercent: number; // Revenue share is private because Web2 platforms do not disclose terms transparently.

    constructor(owner: string) { // Constructor establishes centralized control at the moment of creation.
        this.owner = owner; // Stores the owner string for later reference.
        this.userData = new Map(); // Initializes empty Map because no users exist yet.
        this.revenueSharePercent = 50.0; // Sets default extraction rate.
    }

    storeData(username: string, data: string): void { // Stores user data in the central Map to simulate a server-side database write.
        this.userData.set(username, data); // Adds the key-value pair to the centralized map.
    }

    getRevenueShare(): number { // Returns revenue share so we can expose extraction rates for comparison.
        return this.revenueSharePercent; // Returns the stored percentage.
    }

    getOwner(): string { // Returns the owner to show who has administrative power over the data.
        return this.owner; // Returns the stored owner string.
    }
}

/// Represents a Web3 peer where users hold their own keys and data.
/// We use a class to encapsulate the peer's local state and cryptographic identity.
class Web3Peer {
    publicKey: string; // Public key acts as pseudonymous identity without requiring email or KYC.
    localData: Map<string, string>; // Local Map simulates self-custody: only this peer controls this data.
    protocolFeePercent: number; // Protocol fee is public and immutable to demonstrate transparent economics.

    constructor(publicKey: string) { // Constructor creates a peer with a cryptographic address, not a username.
        this.publicKey = publicKey; // Stores the public key string for later reference.
        this.localData = new Map(); // Initializes empty Map because the peer starts with no data.
        this.protocolFeePercent = 2.0; // Sets minimal transparent fee.
    }

    storeData(data: string): string { // Stores data locally under a SHA-256 hash to demonstrate content-addressable storage.
        const hash = createHash("sha256").update(data).digest("hex").slice(0, 16); // We hash the content so the key is derived from the data itself, proving integrity.
        this.localData.set(hash, data); // Adds the hash-data pair to the local map.
        return hash; // Returns the hash so the caller can verify the content later.
    }

    getProtocolFee(): number { // Returns the low protocol fee for direct numeric comparison with Web2.
        return this.protocolFeePercent; // Returns the stored percentage.
    }
}

const web2 = new Web2Platform("BigTech Inc"); // Instantiate a Web2 platform to represent a typical cloud service.
web2.storeData("alice", "alice's photos and messages"); // Store user data centrally so we can later contrast ownership models.

const web3Peer = new Web3Peer("0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B"); // Instantiate a Web3 peer with a hexadecimal address string.
const contentHash = web3Peer.storeData("alice's encrypted content"); // Store data locally and capture the content hash for verification.

console.log(`Web2 platform takes ${web2.getRevenueShare()}% of revenue`); // Log revenue extraction so the economic difference is immediately visible.
console.log(`Web3 protocol takes ${web3Peer.getProtocolFee()}% of revenue`); // Log the low Web3 fee for direct contrast.

const savingsMultiplier = web2.getRevenueShare() / web3Peer.getProtocolFee(); // Compute savings multiplier to quantify the benefit for creators.
console.log(`Creator keeps ${savingsMultiplier.toFixed(1)}x more in Web3`); // Output the computed savings multiplier.

console.log(`Web2 data controlled by: ${web2.getOwner()}`); // Verify ownership models by printing the controlling entity for each system.
console.log(`Web3 data controlled by: ${web3Peer.publicKey}`); // Print the Web3 peer's public key to show user control.
console.log(`Content hash (integrity proof): ${contentHash}`); // Print the content hash to demonstrate content-addressing.

File: src_web3/phase0/web3_vs_web2_concept.rs
use std::collections::HashMap; // We import HashMap to store key-value data for Web2 and Web3 models.

// PHASE 0: Web3 vs Web2 Concept in Rust
// This file demonstrates why Web3 architectures differ from Web2 using concrete data structures.

/// Represents a traditional Web2 platform where one company owns and controls all user data.
/// We use a struct to group related fields because Web2 platforms have centralized state.
struct Web2Platform {
    owner: String, // The single owner has unilateral power to modify any state; this is the root of trust issues.
    user_data: HashMap<String, String>, // A centralized HashMap stores all user data because there is only one database.
    revenue_share_percent: f64, // The company keeps 100% of revenue distribution logic hidden from users.
}

impl Web2Platform {
    fn new(owner: &str) -> Self { // Creates a new Web2 platform to show how power concentrates at initialization.
        Self {
            owner: owner.to_string(), // Converts the input string to an owned String for storage.
            user_data: HashMap::new(), // Initializes empty HashMap because no users exist yet.
            revenue_share_percent: 50.0, // Users typically keep less than half.
        }
    }

    fn store_data(&mut self, username: &str, data: &str) { // Stores data under a username to simulate a platform writing to its central database.
        self.user_data.insert(username.to_string(), data.to_string()); // Inserts a cloned key-value pair into the centralized map.
    }

    fn get_revenue_share(&self) -> f64 { // Returns revenue share so we can expose the platform's extraction rate numerically.
        self.revenue_share_percent // Returns the stored percentage for comparison with Web3.
    }
}

/// Represents a Web3 protocol where users control their own data through cryptographic keys.
/// We model a peer with its own state to reflect that no central database exists.
struct Web3Peer {
    public_key: String, // The public key serves as identity because there are no usernames or email registrations.
    local_data: HashMap<String, String>, // Each peer stores only its own data locally, mirroring self-custody in real wallets.
    protocol_fee_percent: f64, // Protocol-level revenue share is transparent and immutable, unlike hidden Web2 terms.
}

impl Web3Peer {
    fn new(public_key: &str) -> Self { // Creates a peer with a cryptographic identity to demonstrate pseudonymous ownership.
        Self {
            public_key: public_key.to_string(), // Converts the input string to an owned String for storage.
            local_data: HashMap::new(), // Initializes empty HashMap because the peer starts with no data.
            protocol_fee_percent: 2.0, // Minimal protocol fee visible to everyone.
        }
    }

    fn store_data(&mut self, content_hash: &str, data: &str) { // Stores data locally under a content hash to simulate content-addressable storage.
        self.local_data.insert(content_hash.to_string(), data.to_string()); // Inserts a cloned key-value pair into the local map.
    }

    fn get_protocol_fee(&self) -> f64 { // Returns the low protocol fee so callers can compare extraction directly with Web2.
        self.protocol_fee_percent // Returns the stored percentage for comparison with Web2.
    }
}

fn main() {
    let mut web2 = Web2Platform::new("BigTech Inc"); // Initialize a Web2 platform to represent a typical social network.
    web2.store_data("alice", "alice's photos and messages"); // Store user data centrally so we can show who controls it later.

    let mut web3_peer = Web3Peer::new("0xAb5801a7D398351b8bE11C439e05C5B3259aeC9B"); // Initialize a Web3 peer using a hexadecimal public key string.
    web3_peer.store_data("QmYwAPJzv5CZsnAzt8auVZRn8dPpcXq", "alice's encrypted content"); // Store data locally under a hash to show content-addressing instead of username lookup.

    println!("Web2 platform takes {:.1}% of revenue", web2.get_revenue_share()); // Print revenue extraction comparison so the numeric advantage is visible.
    println!("Web3 protocol takes {:.1}% of revenue", web3_peer.get_protocol_fee()); // Print the low Web3 fee for direct contrast.

    let savings_multiplier = web2.get_revenue_share() / web3_peer.get_protocol_fee(); // Calculate the multiplier of savings to make the benefit concrete.
    println!("Creator keeps {:.1}x more in Web3", savings_multiplier); // Output the computed savings multiplier.

    println!("Web2 data controlled by: {}", web2.owner); // Verify data ownership: in Web2 the platform owns the mapping.
    println!("Web3 data controlled by: {}", web3_peer.public_key); // In Web3 the peer holds the keys.
}
